Source/Dafny/Compiler-java.cs

```diff
@@ -608,12 +604,20 @@ protected class ClassWriter : IClassWriter {
       return TypeName(type, wr, tok, boxed: true);
     }
 
-    private string QExtendsBoxedTypeName(Type type, TextWriter wr, Bpl.IToken tok) {
-      if (type.IsTraitType) {
-        return "? extends " + TypeName(type, wr, tok, boxed: true);
-      } else {
-        return TypeName(type, wr, tok, boxed: true);
+    private string ActualTypeArgument(Type type, TypeParameter.TPVariance variance, TextWriter wr, Bpl.IToken tok) {
+      Contract.Requires(type != null);
+      Contract.Requires(wr != null);
+      Contract.Requires(tok != null);
+      var typeName = BoxedTypeName(type, wr, tok);
+      if (variance == TypeParameter.TPVariance.Co) {
+        // TODO: possible change: if "type" has no subtypes in Java, then return just "typeName"
```

Comment:
I don't know of any downside to using a redundant wildcard in generated code, so I wouldn't spend time trying to detect cases where we're 100% sure there are no subtypes possible.